File: DynamicProgramming/MaxSubarray.py
class Solution:
    # @param {integer[]} nums
    # @return {integer}
    def maxSubArray(self, nums):
        # initialise dp subarray
        dp = [0]*len(nums)
        # intialise base case for DP, first in array is best
        dp[0] = nums[0]
        
        # DP/Compare step
        for i in range(1, len(nums)):
            dp[i] = max(nums[i], nums[i] + dp[i-1])
        return max(dp)

File: DynamicProgramming/MaxSubarray.java
/*
Time O(n), Space O(n)
One Scan
*/

public class Solution {
    public int maxSubArray(int[] nums) {
        //Initialise a DP array 
        int[] dp = new int[nums.length];
        // Arrays.fill(array, 0); (A default value of 0 for arrays)
        
        // First element is best
        dp[0] = nums[0];
        
        int maxSum = dp[0];
        
        // Compare/Dp step
        for (int i = 1; i < nums.length; i++) {
            dp[i] = Math.max(nums[i], nums[i] + dp[i-1]);
            maxSum = Math.max(dp[i], maxSum);
        }
        return maxSum;
    }
}
